### DISCUSSION TITLE
When to use fuzz and/or concrete tests

### DISCUSSION THREAD

**Discussion by AmadiMichael** [alexfertel/bulloak#49] [General]
**Question**: On unit testing, was wondering how to decide if to use concrete testing, fuzz testing or both. Are there any set of questions/checks to ask oneself about the assumptions and expectations of a functionality when deciding if it should have fuzz or/and concrete test?

**Comment by PaulRBerg** [comment]
You typically want to do both.

You want concrete tests to prove that the contracts work with specific values, and you want fuzz tests to test the whole gamut of input values. Concrete tests are also useful for debugging.

However, if you are constrained by time, and you only have time for one category, I'd go with fuzz tests.

Related: see the [test categorization](https://prberg.com/presentations/ethcc-2023/#/8) I have introduced in my EthCC 2023 talk:

<img width="500" alt="SCR-20231121-jkcj" src="https://github.com/alexfertel/bulloak/assets/8782666/ee31292e-5c63-4df8-bff2-7fb960461ddb">
